slashing/program/src/duplicate_block_proof.rs

```diff
@@ -0,0 +1,830 @@
+        msg!(
+            "Verifying duplicate proof for {} slot {}",
+            node_pubkey,
+            slot
+        );
+        // TODO: verify through instruction inspection that the shreds were sigverified
+        // earlier in this transaction.
+        // Ed25519 Singature verification is performed on the merkle root:
+        // node_pubkey.verify_strict(merkle_root, signature).
+        // We will verify that the pubkey merkle root and signature match the shred and
+        // that the verification was successful.
+        let shred1 = Shred::new_from_payload(self.shred1)?;
+        let shred2 = Shred::new_from_payload(self.shred2)?;
+        check_shreds(slot, &shred1, &shred2)
+    }
+}
+
+/// Check that `shred1` and `shred2` indicate a valid duplicate proof
+///     - Must be for the same slot `slot`
+///     - Must be for the same shred version
+///     - Must have a merkle root conflict, otherwise `shred1` and `shred2` must
+///       have the same `shred_type`
+///     - If `shred1` and `shred2` share the same index they must be not have
+///       equal payloads excluding the retransmitter signature
+///     - If `shred1` and `shred2` do not share the same index and are data
+///       shreds verify that they indicate an index conflict. One of them must
+///       be the LAST_SHRED_IN_SLOT, however the other shred must have a higher
+///       index.
+///     - If `shred1` and `shred2` do not share the same index and are coding
+///       shreds verify that they have conflicting erasure metas
+fn check_shreds(slot: Slot, shred1: &Shred, shred2: &Shred) -> Result<(), SlashingError> {
+    if shred1.slot()? != slot {
+        msg!(
+            "Invalid proof for different slots {} vs {}",
+            shred1.slot()?,
+            slot,
+        );
+        return Err(SlashingError::SlotMismatch);
+    }
+
+    if shred2.slot()? != slot {
+        msg!(
+            "Invalid proof for different slots {} vs {}",
+            shred1.slot()?,
+            slot,
+        );
+        return Err(SlashingError::SlotMismatch);
+    }
+
+    if shred1.version()? != shred2.version()? {
+        msg!(
+            "Invalid proof for different shred versions {} vs {}",
+            shred1.version()?,
+            shred2.version()?,
+        );
+        return Err(SlashingError::InvalidShredVersion);
+    }
+
+    // Merkle root conflict check
+    if shred1.fec_set_index()? == shred2.fec_set_index()?
+        && shred1.merkle_root()? != shred2.merkle_root()?
+    {
+        // Legacy shreds are discarded by validators and already filtered out
+        // above during proof deserialization, so any valid proof should have
+        // merkle roots.
+        msg!(
+            "Valid merkle root conflict for fec set {}, {:?} vs {:?}",
+            shred1.fec_set_index()?,
+            shred1.merkle_root()?,
+            shred2.merkle_root()?
+        );
+        return Ok(());
+    }
+
+    // Overlapping fec set check
```

Comment:
Overlapping fec sets are net yet implemented on the detection side however I will work on it soon https://github.com/solana-labs/solana/issues/33037